# Numpy, lists and tuples practicum

### Start by importing numpy:

In [1]:
import numpy as np

### First create an array x from 100 to 1,000 with a step of 2:

In [2]:
x = np.arange( 100, 1000, 2 )

### then create an array y from 1,000 to 2,800 with a step of 4:

In [3]:
y = np.arange( 1000, 2800, 4 )

### check the shape of x and y:

In [4]:
print( x.shape )
print( y.shape )

(450,)
(450,)


### get the length of x:

In [5]:
print( len( x ) )

450


### Create array z equal to the difference of y minus x 

In [6]:
z = y - x

### print it:

In [7]:
#print( z )

[ 900  902  904  906  908  910  912  914  916  918  920  922  924  926
  928  930  932  934  936  938  940  942  944  946  948  950  952  954
  956  958  960  962  964  966  968  970  972  974  976  978  980  982
  984  986  988  990  992  994  996  998 1000 1002 1004 1006 1008 1010
 1012 1014 1016 1018 1020 1022 1024 1026 1028 1030 1032 1034 1036 1038
 1040 1042 1044 1046 1048 1050 1052 1054 1056 1058 1060 1062 1064 1066
 1068 1070 1072 1074 1076 1078 1080 1082 1084 1086 1088 1090 1092 1094
 1096 1098 1100 1102 1104 1106 1108 1110 1112 1114 1116 1118 1120 1122
 1124 1126 1128 1130 1132 1134 1136 1138 1140 1142 1144 1146 1148 1150
 1152 1154 1156 1158 1160 1162 1164 1166 1168 1170 1172 1174 1176 1178
 1180 1182 1184 1186 1188 1190 1192 1194 1196 1198 1200 1202 1204 1206
 1208 1210 1212 1214 1216 1218 1220 1222 1224 1226 1228 1230 1232 1234
 1236 1238 1240 1242 1244 1246 1248 1250 1252 1254 1256 1258 1260 1262
 1264 1266 1268 1270 1272 1274 1276 1278 1280 1282 1284 1286 1288 1290
 1292 

### multiply x by 2, subtract 100 from y and add all this togeter ( (x*2) + (y-100)); assign to variable test1 

In [8]:
test1 = ( x*2 ) + ( y-100 )

### calculate and print the min of test1, its max and its mean:

In [9]:
print( np.min( test1 ), np.max( test1 ), np.mean( test1 ) )

1100 4692 2896.0


### print the mean of the base 10 logarithm of x

In [10]:
print( np.mean( np.log10( x ) ) )

2.675704070469898


### 1. Create a numpy array x that goes from 1 up to and including 100 (last element must be 100.0) with a step of 0.1. 
- Verify its type using the type command and its size using the .shape command.  
- Print x. 
- Print its maximum, minimum and mean. 
- Add 5 to x and print it. 
- Print an array whose values are 10 times larger than x.

In [11]:
x = np.arange( 1, 100.1, 0.1 )

In [15]:
# print( x )

In [16]:
print( np.max( x ), np.min( x ), np.mean( x ) )

100.00000000000009 1.0 50.500000000000036


In [18]:
# print( x + 5 )

In [20]:
x10 =  x * 10 

### 2. Create an array y that is the base 10 logarithm of x. 
- Print y. 
- Print the 10th to 24th elements of x and y. 
- Create an array z that is the same as y. 
- Change the 30th to 140th elements of z to be the 1/10 of their current value and the 500th to 600th elements of z to be 10 times larger than their current value. 
- Print z from the 600 element to the 100 element going backwards with a step of 50 (so the 600th, 550th, 500th etc element).   

In [21]:
y = np.log10( x )

In [22]:
print( x[ 9: 24 ]  )
print( y[ 9: 24 ]  )

[1.9 2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.  3.1 3.2 3.3]
[0.2787536  0.30103    0.32221929 0.34242268 0.36172784 0.38021124
 0.39794001 0.41497335 0.43136376 0.44715803 0.462398   0.47712125
 0.49136169 0.50514998 0.51851394]


In [31]:
z = y 

In [32]:
len(z)

991

In [33]:
z[ 29: 140] = 0.1* z[ 29: 140]

In [35]:
z[ 499: 600 ] = 10* z[ 499: 600 ]

In [36]:
print( z[ 600: 100: -50  ] )

[ 17.85329835 174.8188027  170.75701761   1.66275783   1.61278386
   1.5563025    1.49136169   1.41497335   1.32221929   1.20411998]


### 3. Create array base1 that goes from 100 to 600 with a step of 2. 
- Create array base2 that goes from 75.5 to 0.5 with a step of -0.3. 

- Assuming that base1 is the one side of a (orthogonal parallelogram) field and base2 the other side, calculate the surface area of all fields (field_surface). 
- How many fields are there in your sample? 
- How many unique surface areas are there in your sample?

In [42]:
base1 = np.arange( 100, 600, 2)
base2 = np.arange( 75.5, 0.5, -0.3 )

In [43]:
print( len( base1) , len( base2) ) 

250 250


In [44]:
field_surface = base1* base2

In [45]:
print( len( field_surface ) )

250


In [47]:
#print( np.unique( field_surface ))
print( len( np.unique( field_surface ) ) )

250


### One of the great features of numpy is that it also allows you to read data in an array in a single command. You do that with np.loadtxt( "name_of_file" ) [see also: https://numpy.org/doc/stable/reference/generated/numpy.loadtxt.html ] or np.genfromtxt("name_of_file") [see also: https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html ]. 

### E.g., get file *my_first_data.dat* from the Webcouses/demos/ . Open it with a text editor to see what it looks like. 
- Read it in an array *all_input_data* with loadtxt or genfromtxt. 
- Check what your array looks like (print it). 
- Print all times each car traveled. 
- Print all distances the cars traveled. 
- Calculate the distance each car traveled **in a single line** and asign it to array *distances_traveled*. 
- Print *distances_traveled*

In [48]:
all_input_data = np.loadtxt( '../all_data/my_first_data.dat' )

In [ ]:
# Car_number  Time [hr]  Speed [mph]

In [49]:
print( all_input_data )

[[  1.    5.2 120.2]
 [  2.    3.1 110. ]
 [  3.    2.4 100.5]
 [  4.    5.1  65.3]
 [  5.    2.4 119.6]]


In [50]:
print( all_input_data[ :, 1 ] )

[5.2 3.1 2.4 5.1 2.4]


In [51]:
distance_traveled =  all_input_data[ :, 1 ] * all_input_data[ :, 2 ] 

In [52]:
print( distance_traveled )

[625.04 341.   241.2  333.03 287.04]


### Now let's try the genfromtxt with data: *test_genfromtxt.dat*. 
- Read data into variable student_data. 
- Print the IDs of the students. 
- Print the average quiz  grade for student 202204. 
- Print the average quiz grade and the IDs of all students that passed the class with a C- or better ( average total of more than 7.0).

In [53]:
student_data = np.genfromtxt( '../all_data/test_genfromtxt.dat' )

In [54]:
print( student_data )

[[2.02201e+05 8.10000e+00 8.40000e+00 8.20000e+00]
 [2.02202e+05 7.50000e+00 7.00000e+00 7.30000e+00]
 [2.02203e+05 7.90000e+00 8.50000e+00 8.40000e+00]
 [2.02204e+05 9.50000e+00 9.70000e+00 9.70000e+00]
 [2.02205e+05 5.30000e+00 6.00000e+00 5.90000e+00]
 [2.02206e+05 5.10000e+00 7.10000e+00 6.80000e+00]]


In [ ]:
#Student ID  Average-Grade-Quiz  Average-Grade-Exam    Average-Total

In [55]:
IDs = student_data[ :, 0 ]
print( IDs )

[202201. 202202. 202203. 202204. 202205. 202206.]


In [58]:
print( student_data[ 3, 1 ])

9.5


In [59]:
student_data[ :, 3 ] > 7 

array([ True,  True,  True,  True, False, False])

In [64]:
print( student_data )

[[2.02201e+05 8.10000e+00 8.40000e+00 8.20000e+00]
 [2.02202e+05 7.50000e+00 7.00000e+00 7.30000e+00]
 [2.02203e+05 7.90000e+00 8.50000e+00 8.40000e+00]
 [2.02204e+05 9.50000e+00 9.70000e+00 9.70000e+00]
 [2.02205e+05 5.30000e+00 6.00000e+00 5.90000e+00]
 [2.02206e+05 5.10000e+00 7.10000e+00 6.80000e+00]]


In [60]:
np.where( student_data[ :, 3 ] > 7  )

(array([0, 1, 2, 3]),)

In [61]:
print( student_data[ np.where( student_data[ :, 3 ] > 7  ) , 1 ])

[[8.1 7.5 7.9 9.5]]


In [62]:
print( student_data[ np.where( student_data[ :, 3 ] > 7  ) , 0 ])

[[202201. 202202. 202203. 202204.]]


In [63]:
print( student_data[  student_data[ :, 3 ] > 7  , 1 ])

[8.1 7.5 7.9 9.5]


### 4. You just opened a new savings account that earns N percent interest per year. The interest that you earn is paid at the end of the year, and is added to the balance of the savings account. Write a *function end_of_year_interest* that takes as input the interest N and a single list *deposits_list* that containts the amount of money deposited into the account from the user at the start of years 1, 2 and 3 (assume this is the only deposit to the account). The function should then calculate the amount in the savings account at the end of years 1, 2 and 3 and print it out. Call the function for an interest rate of 2% and deposits of 100, 300 and 500. 

- Discuss: what should the input and output of your function be?
- What does a good docstring need to include?
- How do we code this up?
- Code!


In [70]:
def end_of_year_interest( N, deposits_list ) :
    """Function that calculates the interest (assumed 4% ) gained per year for 3 years and 
    prints the total balance of savings account at the end of each year. 
    Input: N (interest in %), deposits_list (list with deposit $ for years 1 to 3) 
    Output: total savings amount for years 1 to 3 """
    
    year_1_interest = deposits_list[ 0 ] * ( N / 100. ) 
    year_1_total = deposits_list[ 0 ] * ( 1 + N / 100. )
    
    year_2_interest = ( year_1_total + deposits_list[ 1 ] ) * ( N / 100. )
    year_2_total = ( year_1_total + deposits_list[ 1 ] ) * ( 1 + N / 100. ) 
    
    year_3_interest = ( year_2_total + deposits_list[ 2 ] ) * ( N / 100. ) 
    year_3_total = ( year_2_total + deposits_list[ 2 ] ) * ( 1 + N / 100. )
    
    print( " total interest per year is: " , year_1_interest, year_2_interest, 
         year_3_interest )
    
    print( " total amount per year is: " , year_1_total, year_2_total, 
         year_3_total )
    


In [71]:
end_of_year_interest(2, [100, 300, 500])

 total interest per year is:  2.0 8.040000000000001 18.2008
 total amount per year is:  102.0 410.04 928.2408


### 5. Create a program that calculates how fast an object moves [ m/s ] when falling to the ground from a height h. Assume that the body free falls with initial speed 0 m/s, so that u$_{final}$ = $\sqrt(2 * g * h)$, where g = 9.8 $m/s^2$ and h is in m. 
### Assume that we have a range of object that fall and the recorded height from which they start are stored in a numpy array  ***heights***. Make array ***heights*** from 3 m to 100m with a step of 2.5 (average height of a floor). 

- How will you calculate the final speed of the bodies falling from the different heights?

- Create a variable u_final that calculates the final speed of the falling object from **all** heights in **one** line. 
- Print it

### 6. Create a list *test_list_1* that contains the numbers 1, 2, 5, 7, 22, 34, 42. 
- Print it. 
- Change 22 to -10 (don't just copy-paste and change it in the series, access the appropriate element of test_list_1 and change it). 
- Print the mutated list. 
- Add elements 96 and 102 to the end of the list. 
- Add 0 to the start. 

### 7. Create a tuple *trip_time* that contains the amount of hours a trip takes. Create a tuple *trip_distance* that contains the distance of a given trip in km. <br> Populate *trip_time* with numbers: 1.5, 3.2, 5.6, 14.3, 26.4 and *trip_distance* with 180, 320, 728, 1573, 3326. Calculate the average speed (remember it is distance/time) of each trip. 
- Can you use plain division of the tuples in a single line to do it? Why/why not?  
- After each calculation store the average speed in a tuple called *trip_average_speed* ("append" each calculation to the previous tuple). Is the tuple you have at the end, the same as at the start? Why/why not? 
- You just realized you made a typo, and the 728 should have been 718! Change the tuple accordingly (and properly) and recalculate and print the new *trip_average_speed* tuple. 


### Make numpy arrays  *np_trip_time* and *np_trip_distance* from the original *trip_time* and *trip_distance*. Calculate the average speed of each trip using the numpy arrays and store it in array *np_trip_average_speed*. What is the benefit in comparison to tuples (or lists?)
- You just realized you made the same typo, and the 728 should have been 718! How do you change the array now?

In [ ]:
### also, yes, you can kind of do it in a single line but not in a way we have seen yet, e.g.:
#from operator import truediv 
#trip_average_test_one_line = tuple( map( truediv, trip_distance , trip_time ) )  
#print( trip_average_test_one_line )

In [ ]:
# now let's change the element; remember tuples are immutable!
#inital was trip_distance = ( 180, 320, 728, 1573, 3326 ) --> need to change trip_distance[ 2 ]


### 8. Open a text editor and save it as *all_my_functions.py*. Copy in there the function *acceleration* from your HW and from this and previous practicums the functions *end_of_year_interest* and *thunderstruck*. Change the names of the functions by adding a *_fromfile* at the end of each. <br> In a single command import *acceleration_fromfile* and *end_of_year_interest_fromfile* and *thunderstruck_fromfile*. Call the latter for the case you hear a thunder 10s after you see the lightning. 

#### Note that if you are using numpy in these functions you will need to start your all_my_functions.py with importing numpy as np.

### 9. The Body Mass Index is one of the available measures of how healthy a person is. It is defined as the ratio of a person's weight in kg, over the square of their height in m. Make a function BMI that uses the height and weight of a person and returns their BMI. 

### Make a list heights that contains the heights of people, and a list weights that containts their weights. Populate them with heights: 1.60, 2.10 and 1.80 and weights 60, 120 and 81.

### Call the function BMI for all three people and print their BMI.

###  10. Create a list student_h that contains lists with the heights of students in feet and inches. A person being 5' 6" would be noted as [5, 6], a person being 5' 10" would be noted as [5, 10]. Populate it with the heights of five students who are : 5’ 1”, 4’ 10”, 6’ 02”, 5’ 07”, and 6’ 11”. Create a function ft_to_m_cm that takes as input a value from student_h and returns each height in a new list [m, cm] (round cm up). Call the function for all values of student_h and print the results with an informative sentence (e.g., Student 1 has a height of 1 m 10 cm )  (Remember that 1 ft ~= 0.3048 m and 1 inch ~= 0.0254 m and 1m has 100cm). 

- Discuss: what should the function's input be? 
- What should its output be?
- What does the function need to calculate? How will it do that?
- Code it up!

In [ ]:
#make student_h list:

In [ ]:
#make your function:


In [ ]:
#calculate the heights in m and cm and print with informative statement: